reps/2023-10-13-accelerator-support.md

```diff
@@ -0,0 +1,399 @@
+
+# TPU
+ScalingConfig(num_workers=4, resources_per_worker={"TPU": 1})
+
+# HPU
+ScalingConfig(num_workers=4, resources_per_worker={"HPU": 1})
+```
+
+In the future, Ray Train will restructure the `ScalingConfig` class, and incorporate the `accelerate_type` parameter in a more structured way.
+
+#### 2. Specify PyTorch Backend for each Accelerator Family
+
+Different accelerator family uses different backend to launch deep learning training. Ray Train will implement a different backend class for each accelerator family to launch distributed groups and set environment variables.
+
+| Accelerator | Communication Backend | Developer APIs | User APIs | 
+| - | - | - | - |
+| GPU      | `nccl`, `gloo` | TorchBackend | `TorchConfig()` |
+| TPU      | `xla[tpu]`      | TorchTPUXLABackend | `TorchConfig("xla[tpu]")` |
+| Trainium | `xla[neuronx]`  | TorchAwsNeuronXLABackend | `TorchConfig("xla[neuronx]")`  |
+| HPU      | `hccl`         | TorchHCCLBackend | `TorchConfig("hccl")` |
+| XPU      | `oneCCL`       | TorchOneCCLBackend | `TorchConfig("ccl")`|
+
+
+Putting things together, users should be able to set up the resources and backend with these two APIs in `TorchTrainer`.
+
+```python
+from ray.train import ScalingConfig
+from ray.train.torch import TorchTrainer, TorchConfig
+
+# TPU Training
+trainer = TorchTrainer(
+    train_func,
+    scaling_config=ScalingConfig(
+        num_workers=4, 
+        resources_per_worker={"TPU": 1}
+    ),
+    torch_config=TorchConfig(backend="xla[tpu]")
+)
+
+# AWS Trainium
+trainer = TorchTrainer(
+    train_func,
+    scaling_config=ScalingConfig(
+        num_workers=4, 
+        resources_per_worker={"neuron_cores": 1}
+    ),
+    torch_config=TorchConfig(backend="xla[neuronx]")
+)
+
+# HPU
+trainer = TorchTrainer(
+    train_func,
+    scaling_config=ScalingConfig(
+        num_workers=4, 
+        resources_per_worker={"HPU": 1}
+    ),
+    torch_config=TorchConfig(backend="hccl")
+)
+```
+
+
+### Ray RLLib
+Currently there is no plan to support other accelerators in RLLib.
+
+### Ray Serve
+Nothing needs to be changed on the serve side.
+
+Users can define a deployment using accelerators via `ray_actor_options`:
+
+```python
+@serve.deployment(ray_actor_options={"num_gpus": 1})
+class Deployment:
+    ...
+
+@serve.deployment(ray_actor_options={"resources": {"TPU": 1}})
+class Deployment:
+    ...
+```
+
+### Ray Data
+Nothing needs to be changed on the data side.
+
+Users can use the corresponding accelerator resources for data operations:
+
+```python
+ds.map_batches(..., resources={"TPU": 1})
+```
+
+### Kuberay
+
+To auto populate `rayStartParams` resources using pod resources, some code is needed.
+See https://github.com/ray-project/kuberay/blob/b7bc7ae8c983160bb700d6b8bb4014dda183ecea/ray-operator/controllers/ray/common/pod.go#L687 as an example of how GPU resource is populated.
+
+### Observability
+
+Currently we emit metrics about the physical usages of GPUs and also show them on Ray dashboard.
```

Comment:
Can we make this as a part of `AcceleratorManager` class? We also need ways to emit metrics. 

We need: acclerator (TPU vs GPU), acclerator_type (A100, H100), utilization & memory. 

1. Dashboard cluster view should display live metrics.
2. Dashboard should emit metrics with device type (with different labels). 
3. ray status should correctly display devices.

Reply:
Yes, it will. I'm trying some prototype here: https://github.com/ray-project/ray/pull/40749.